src/imageops/filter_1d.rs

```diff
@@ -0,0 +1,879 @@
+            k3 = mla(k3, fw[3].as_() + bw[3].as_(), coeff);
+        }
+
+        dst[0] = k0.to_();
+        dst[1] = k1.to_();
+        dst[2] = k2.to_();
+        dst[3] = k3.to_();
+    }
+
+    let dzx = dst_row.chunks_exact_mut(4).len() * 4;
+    let remainder = dst_row.chunks_exact_mut(4).into_remainder();
+
+    for (x, dst) in remainder.iter_mut().enumerate() {
+        let v_cx = x + dzx;
+        let src = &src[v_cx..];
+
+        let mut k0 = src[half_len * N].as_() * hc;
+
+        for (i, &coeff) in scanned_kernel.iter().take(half_len).enumerate() {
+            let other_side = length - i - 1;
+            let fw = &src[(i * N)..(i * N) + 1];
+            let bw = &src[(other_side * N)..(other_side * N) + 1];
+            k0 = mla(k0, fw[0].as_() + bw[0].as_(), coeff);
+        }
+
+        *dst = k0.to_();
+    }
+}
+
+trait KernelTransformer<F, I> {
+    fn transform(input: F) -> I;
+}
+
+impl KernelTransformer<f32, u32> for u8 {
+    fn transform(input: f32) -> u32 {
+        const SCALE: f32 = (1 << Q0_15) as f32;
+        (input * SCALE).min(((1u32 << Q0_15) - 1) as f32).max(0.) as u32
+    }
+}
+
+impl KernelTransformer<f32, f32> for f32 {
+    fn transform(input: f32) -> f32 {
+        input
+    }
+}
+
+impl KernelTransformer<f32, f32> for u16 {
+    fn transform(input: f32) -> f32 {
+        input
+    }
+}
+
+/// Removes meaningless values from kernel preserving symmetry
+fn prepare_symmetric_kernel<I: Copy + PartialEq + 'static>(kernel: &[I]) -> Vec<I>
+where
+    i32: AsPrimitive<I>,
+{
+    let zeros: I = 0i32.as_();
+    let mut new_kernel = kernel.to_vec();
+    while new_kernel.len() > 2
+        && (new_kernel.last().unwrap().eq(&zeros) && new_kernel.first().unwrap().eq(&zeros))
+    {
+        new_kernel.remove(0);
+        new_kernel.remove(new_kernel.len() - 1);
+    }
+
+    new_kernel
+}
+
+const RING_QUEUE_CIRCULAR_CUTOFF: usize = 55;
+
+fn filter_1d_ring_queue<T, F, I, const N: usize>(
+    image: &[T],
+    destination: &mut [T],
+    image_size: FilterImageSize,
+    row_kernel: &[F],
+    column_kernel: &[F],
+) -> Result<(), ImageError>
+where
+    T: Copy + AsPrimitive<F> + Default + Send + Sync + KernelTransformer<F, I> + AsPrimitive<I>,
+    F: Default + 'static + Copy,
+    I: ToStorage<T>
+        + Mul<I, Output = I>
+        + Add<I, Output = I>
+        + MulAdd<I, Output = I>
+        + Send
+        + Sync
+        + PartialEq
+        + Default
+        + 'static
+        + Copy,
+    i32: AsPrimitive<F> + AsPrimitive<I>,
+    f64: AsPrimitive<T>,
+{
+    if image.len() != image_size.width * image_size.height * N {
+        return Err(ImageError::Limits(LimitError::from_kind(
+            LimitErrorKind::DimensionError,
+        )));
+    }
+    if destination.len() != image_size.width * image_size.height * N {
+        return Err(ImageError::Limits(LimitError::from_kind(
+            LimitErrorKind::DimensionError,
+        )));
+    }
```

Comment:
Both of these are overflow sensitive computations. And as a post-condition we have that both buffers are equal length which should be interesting to readers of certain copies below.